src/routes/logic.js

```diff
@@ -0,0 +1,185 @@
+/* eslint-disable */
+import {fsm, util} from 'message-fsm';
+import launch_fsm from './launch.fsm.js';
+import yaml from 'js-yaml';
+
+async function postData(url = ``, data = {}) {
+  // Default options are marked with *
+    var response = await fetch(url, {
+        method: "POST", // *GET, POST, PUT, DELETE, etc.
+        mode: "cors", // no-cors, cors, *same-origin
+        cache: "no-cache", // *default, no-cache, reload, force-cache, only-if-cached
+        credentials: "same-origin", // include, *same-origin, omit
+        headers: {
+            "Content-Type": "application/json; charset=utf-8",
+            // "Content-Type": "application/x-www-form-urlencoded",
+        },
+        redirect: "follow", // manual, *follow, error
+        referrer: "no-referrer", // no-referrer, *client
+        body: JSON.stringify(data), // body data type must match "Content-Type" header
+    })
+	  return await response.json();
+}
+
+class Controller {
+
+  constructor(app) {
+    window.controller = this;
+    this.app = app;
+    this.protocol = "https://"
+    this.server = "nginx-find-it-fix-it.5a9f.insights-dev.openshiftapps.com"
+    this.launch = this.launch.bind(this);
+    this.cancel = this.cancel.bind(this);
+    this.get_api_object = this.get_api_object.bind(this);
+    this.get_api_list = this.get_api_list.bind(this);
+    this.get_api_url = this.get_api_url.bind(this);
+    this.poll_log = this.poll_log.bind(this);
+    this.poll_status = this.poll_status.bind(this);
+		this.plan_id = 1;
+		this.inventory_id = 1;
+		this.key_id = 1;
+		this.worker_id = 1;
+    this.trace_order_seq = util.natural_numbers(0);
+    this.controller = new fsm.FSMController(this, 'launch_fsm', launch_fsm.Start, this);
+    this.channel = new fsm.Channel(null, this.controller, this);
+    this.launch_enabled = true;
+    this.cancel_enabled = false;
+    this.hosts = [];
+    this.playbook = {name: "", contents: ""};
+    this.doc = [];
+    this.playbook_run_id = null;
+    this.plays = [];
+    this.log = [];
+    this.tasks_by_host = {};
+  }
+
+  async init() {
+    this.hosts = await this.get_api_list('host', 'inventory=' + this.inventory_id);
+    this.playbook = await this.get_api_object('playbook', 'plan_id=' + this.plan_id);
+    this.doc = yaml.safeLoad(this.playbook.contents);
+    var i = 0;
+    for (i = 0; i < this.doc.length; i++) {
+      if (this.doc[i].name !== undefined) {
+        this.plays.push(this.doc[i].name);
+      }
+    }
+    this.app.setState({});
+    for(i = 0; i < this.hosts.length; i++) {
+      this.tasks_by_host[this.hosts[i].host_id] = [];
+    }
+  }
+
+  send_trace_message(message) {
+    console.log(message)
+  }
+
+	get_api_url(name) {
+    return this.protocol + this.server + '/insights_integration/api/' + name + '/';
```

Comment:
`urijs` is already a dependency. You may want to use that to make URL building more robust

Reply:
Thanks for the tip